src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs

```diff
@@ -628,6 +628,51 @@ public async Task<ContentValidateResult> ValidateAsync(ContentItem contentItem)
             return validateContext.ContentValidateResult;
         }
 
+        public async Task<ContentValidateResult> RestoreAsync(ContentItem contentItem)
+        {
+            var validateContext = new ValidateContentContext(contentItem);
+
+            await Handlers.InvokeAsync((handler, context) => handler.ValidatingAsync(context), validateContext, _logger);
+
+            await ReversedHandlers.InvokeAsync((handler, context) => handler.ValidatedAsync(context), validateContext, _logger);
+
+            if (!validateContext.ContentValidateResult.Succeeded)
+            {
+                await _session.CancelAsync();
+                return validateContext.ContentValidateResult;
+            }
+
+            // So that a new record will be created.
+            contentItem.Id = 0;
+
+            // So that a new version id will be generated.
+            contentItem.ContentItemVersionId = "";
```

Comment:
@deanmarcussen 

Yes good to move this method in the `DefaultContentManager`, i thought about it but i was hesitating, i don't remember the details but because of how the audit trail works we use the content item embedded in the audit event, and here **we create a new version**, and i was thinking that we may need here another `RestoreAsync()` that would be more oriented on existing version ids, and that would just restore an existing version (maybe by still making it the new last one) but **without creating a new record and then without using a new version id**. And maybe the parameter would be just a `contentItemVersionId`.

---

Good to add validate handlers, i also thought about this but i thought that a given version was already validated (but yes in an old life), and knowing that when a new last draft is created from it we will need to validate it again to edit it and then to save it again or publish it. The idea being to not block the restore of a given version and let the user edit it and then have a chance to update it if the validation don't pass again.

Hmm but just thought about another scenario, if we restore a version that becomes the new last draft and then we do a `publish draft` action on it without editing it, no further validation would be performed, so maybe better to validate it on restore, but i think that you got my idea.

Reply:
> @deanmarcussen
> 
> Yes good to move this method in the `DefaultContentManager`, i thought about it but i was hesitating, i don't remember the details but because of how the audit trail works we use the content item embedded in the audit event, and here **we create a new version**, and i was thinking that we may need here another `RestoreAsync()` that would be more oriented on existing version ids, and that would just restore an existing version (maybe by still making it the new last one) but **without creating a new record and then without using a new version id**. And maybe the parameter would be just a `contentItemVersionId`.

Yes, had a little chat about it at the last meeting, and everyone seemed to agree it made sense to move them there, even if audit trail is the only ui based way to restore currently.

I agree also we will need a `RestoreAsync` that would take an existing verion id, and flip the `Latest` / `Published`.

I've got a branch tucked away here https://github.com/OrchardCMS/OrchardCore/tree/deanmarcussen/audittrail-historyview which has what was a very quick view of content item versions, regardless of what is in the audit trail / or potentially what has been trimmed. I'm thinking to add the other `RestoreAsync` to that feature, after this feature, as I don't want to block it anymore!

> Good to add validate handlers, i also thought about this but i thought that a given version was already validated (but yes in an old life), and knowing that when a new last draft is created from it we will need to validate it again to edit it and then to save it again or publish it. The idea being to not block the restore of a given version and let the user edit it and then have a chance to update it if the validation don't pass again.

In this case I just kept what @domonkosgabor already had, which I think was good. the `ValidateAsync` will just check that the Api is safe, i.e. the route is not used somewhere else, or the alias. 
Because users can be using those indexes / the slug providers, so unicity is important. I will check, but I think we are returning the error message, so it would give the user the opportunity to go find the other content item sharing the alias / route, to change it (if they _must_ restore this item).

So I thought it was probably best to block the restore of a content item in this case, as it will actually throw the system off. What it won't do, and we will also have to look at later, is when a content type has been removed, the restored version will probably throw display handler exceptions. We have an issue tracking that though, and in the short term, it might just have to be buyer beware. Knowing if they restore a bad type, they can unrestore it, back to the working version.

That being said though, we could try catch it here, and prevent it, but I'd rather get a better solution in for the actual issue, around missing content types.

And also, of course, an audit trail for content type definitions ;) 

Another question:
I have moved a number of the features around a little and there was a lot of places you had put `RequireFeatures` attributes on things like the providers etc. Are they necesary, or? 
(I will just have to tweak them to use `Features` )

Reply:
Another question:

I am still wondering if we should be calling the `UpdateHandlers` here as well, as `Create` - I know you and @Skrypt did some work around those extension methods a while ago.

Reply:
@deanmarcussen Okay cool then, just wanted you to be aware of what i was thinking at some point.

So, yes if we call validating handlers we could also call `UpdateAsync()` that just save the item but also calls updating handlers. At some point i thought about just using the `UpdateValidateAndCreateAsync()` extension that @Skrypt created, but here we are creating a new draft but not creating a new content item related to a new `contentItemId`.

There is no `UpdateAndValidateAsync()` ;)

Note: Here we still need to do specific things around the item id and version id of the embedded item retrieved from the audit event, and around the current latest version.

But yes, now we could also call `UpdateAsync()`, but i think before validating. At some point we didn't call updating / updated handlers to not fire other audit trail events, but as i remember it is checked in the handlers, that said you will need to call the `Restoring` handlers first, before the updating / validating handlers, and restored ones at the end.

Finally, now maybe better to first check **if the updating and the validation don't fail**, this before the `Remove an existing draft but keep an existing published version`, hmm but if the session is canceled, normally that's okay.

Reply:
cool, I didn't get to reviewing this properly today, but think it all makes sense @jtkech . Think `Update` should probably be called before `Validate` as you say, as it gives the opportunity to regen an autoroute if required. (so actually fixes the issue of bringing an incorrect route in, as it will prefix it -1 if needed.) Need to check and test
And `Restoring` first makes sense to, other opportunity to do anything required.

If you look tonight you will see I moved the Providers / Descriptors to Options - it was too much the constant building of descriptors all the time. Should be ok the change, but appreciate your opinion.

**Another question I need to check with you on**
I have moved a number of the features around a little and there was a lot of places you had put RequireFeatures attributes on things like the providers etc. Are they necesary, or?
(I will just have to tweak them to use Features )

Reply:
@deanmarcussen 

Okay, i saw the new options and how string localizers are lazily resolved, cool.

About RequireFeatures attributes, we should keep those that are used for startups that are not tied to a separate feature but that require other features. Then yes you can remove those that are related to services that are registered in these already decorated statups. But still needed for specific Migrations and Controllers because of the usage of `_typeFeatureProvider`, here for controllers:

https://github.com/OrchardCMS/OrchardCore/blob/8ed77d612917dda3816a2f28a3de33fdb7df6dfd/src/OrchardCore/OrchardCore.Mvc.Core/ModularApplicationModelProvider.cs#L46

And here for migrations:

https://github.com/OrchardCMS/OrchardCore/blob/8ed77d612917dda3816a2f28a3de33fdb7df6dfd/src/OrchardCore/OrchardCore.Data.YesSql/Migration/DataMigrationManager.cs#L89

Reply:
@jtkech ok content manager events sorted (hopefully). Would appreciate a second look if you have time.

I think everything else is ready to go as well, so if you are ok to cast your eye over it @domonkosgabor and @jtkech that would be great.

Reply:
@deanmarcussen

Okay LGTM i will approve, just waiting a little for @domonkosgabor 's feedback.